parser/prism/Factory.cc

```diff
@@ -0,0 +1,502 @@
+    ENFORCE(returnType, "Return type is null");
+
+    pm_node_t *builder = T(loc);
+
+    builder = Send0(loc, builder, "proc"sv);
+
+    if (args != nullptr) {
+        builder = Send1(loc, builder, "params"sv, args);
+    }
+
+    return Send1(loc, builder, "returns"sv, returnType);
+}
+
+pm_node_t *Factory::TProcVoid(core::LocOffsets loc, pm_node_t *args) const {
+    pm_node_t *builder = T(loc);
+
+    builder = Send0(loc, builder, "proc"sv);
+
+    if (args != nullptr) {
+        builder = Send1(loc, builder, "params"sv, args);
+    }
+
+    return Send0(loc, builder, "void"sv);
+}
+
+pm_node_t *Factory::Send2(core::LocOffsets loc, pm_node_t *receiver, string_view method, pm_node_t *arg1,
+                          pm_node_t *arg2) const {
+    ENFORCE(receiver && !method.empty() && arg1 && arg2, "Receiver or method or arguments are null");
+
+    pm_constant_id_t methodId = addConstantToPool(method);
+    vector<pm_node_t *> args = {arg1, arg2};
+    pm_arguments_node_t *arguments = createArgumentsNode(args, parser.convertLocOffsets(loc));
+
+    pm_location_t fullLoc = parser.convertLocOffsets(loc);
+    pm_location_t tinyLoc = parser.convertLocOffsets(loc.copyWithZeroLength());
+
+    return up_cast(createSendNode(receiver, methodId, up_cast(arguments), tinyLoc, fullLoc, tinyLoc));
+}
+
+pm_node_t *Factory::TLet(core::LocOffsets loc, pm_node_t *value, pm_node_t *type) const {
+    ENFORCE(value && type, "Value or type is null");
+    return Send2(loc, T(loc), "let"sv, value, type);
+}
+
+pm_node_t *Factory::TCast(core::LocOffsets loc, pm_node_t *value, pm_node_t *type) const {
+    ENFORCE(value && type, "Value or type is null");
+    return Send2(loc, T(loc), "cast"sv, value, type);
+}
+
+pm_node_t *Factory::TMust(core::LocOffsets loc, pm_node_t *value) const {
+    ENFORCE(value, "Value is null");
+    return Send1(loc, T(loc), "must"sv, value);
+}
+
+pm_node_t *Factory::TUnsafe(core::LocOffsets loc, pm_node_t *value) const {
+    ENFORCE(value, "Value is null");
+    return Send1(loc, T(loc), "unsafe"sv, value);
+}
+
+pm_node_t *Factory::TAbsurd(core::LocOffsets loc, pm_node_t *value) const {
+    ENFORCE(value, "Value is null");
+    return Send1(loc, T(loc), "absurd"sv, value);
+}
+pm_node_t *Factory::TBindSelf(core::LocOffsets loc, pm_node_t *type) const {
+    ENFORCE(type, "Type is null");
+
+    return Send2(loc, T(loc), "bind"sv, Self(loc), type);
+}
+
+pm_node_t *Factory::TTypeAlias(core::LocOffsets loc, pm_node_t *type) const {
+    ENFORCE(type, "Type is null");
+
+    pm_node_t *send = Send0(loc, T(loc), "type_alias"sv);
+
+    pm_statements_node_t *stmts = allocateNode<pm_statements_node_t>();
+    *stmts = (pm_statements_node_t){.base = initializeBaseNode(PM_STATEMENTS_NODE, parser.convertLocOffsets(loc)),
+                                    .body = {.size = 0, .capacity = 0, .nodes = nullptr}};
+    pm_node_list_append(&stmts->body, type);
```

Comment:
This can reuse the factory function.

Once everything takes a span, we can use a regular stack-allocated array at the input. Something like:

```suggestion
    auto *stmts = StatementsNode(parser.convertLocOffsets(loc), std::array{type})
```

Then you can delete the `up_cast` below